### DISCUSSION TITLE
Remove files from the index only

### DISCUSSION THREAD

**Discussion by RaddadZ** [photoprism/photoprism#2798] [Q&A] [answered]
Hi there, 

**TL;DR**
I couldn't find a way to reindex specific photos other than by deleting 5 entries in the database and files from 3 different directories in `storage` for every photo. is there any other way to do it?

I have a VM that runs docker that I wanted to run Photoprism on it. The container runs well and smoothly but when I tried to index around 20 GB of raw images (~30mB each) the VM shut down.
When restarted the VM resumed the index operation it finished successfully but with ~20 or so photos not showing thumbs and/or the photo itself (other metadata shows up) and they are all stored perfectly in the database, not even with an error 

It's fair since the VM shut down while indexing and after a brief search in the docs it points to doing a full rescan, which is costly, and I know that the VM will shut down again and the problem will reemerge (since it's hardware related I guessed).
So I was in need of a way to reindex files without removing them physically or doing a full rescan; which was to try and remove everything related to these files in the database or in the storage directory. 

After much research, I couldn't find a clear map of what the "index" operation exactly does, until I checked the code, 
It turns out it stores data related to the files in 5 different tables, and in `storage/cache/json`, `storage/cache/thumbnails`, and `storage/sidecar` directories, so a pseudo code to remove these related rows and files is something like this:
```
@filenames = ('subfolder/filename.ext')
@photo_ids = select photo_id from files where file_name in (@filenames)
@file_hashs = select file_hash from files where file_name in (@filenames)

# sql queries
delete from files where photo_id in (@photo_ids);
delete from photos where id in (@photo_ids);
delete from details where photo_id in (@photo_ids);
delete from photos_keywords where photo_id in (@photo_ids);
delete from photos_labels where photo_id in (@photo_ids);

# terminal
foreach(@hash in @file_hashs)
    rm ./storage/cache/json/@hash[0:1]/@hash[1:2]/@hash[2:3]/@hash*
    rm ./storage/cache/thumbnails/@hash[0:1]/@hash[1:2]/@hash[2:3]/@hash*

foreach(@filename in @filenames)
    rm ./storage/sidecar/@remove_ext(@filename)*
```

my Q is, is there any other way to remove files from the index and only the index, not physically? 
or is there any other way to force indexing specific files in the cli or the UI?

**Comment by lastzero** [comment, marked as answer]
You can (re-)index specific folders only, but not individual files, because they can belong together, e.g. a JPEG and a RAW. Assuming you force the indexing of a RAW without the JPEG, you will have exactly the problem (no JPEG, no preview) you actually wanted to fix. Whether your directory structure is suitable for partial indexing or whether you put all the files in the same directory is beyond our control and knowledge.

**Reply by RaddadZ** [reply]
Thanks for the clarification. I'll check the docs and the code again to get a better grasp on the subject of files belonging to each other.
But I'm assuming the app can know which files belong to each other, hence (re-)indexing these files should be possible. Or maybe it should be possible to point out what is missing (RAW file, or JPEG file).

I think if we could have a way to (un)index photos without the need to physically remove them would be nice.

Of course, if I had read the docs carefully I would've organized my files in a different way. letting me index them by folder and with better "checkpoints" for unexpected failure.

I think I got my answer, thanks again.
